specs/agents/tracing-instrumentation-messaging.md

```diff
@@ -0,0 +1,103 @@
+event for a message handling flow. Capturing polling spans is also mostly straightforward. 
+For polling-based transactions, our goal is to capture the message handling flow, which typically *starts after the polling action exits*, 
+returning a message. This may be tricky if the handling flow is not implemented within a well defined API. Use other agent implementation
+as reference.
+The agent should not create a transaction based on polling APIs if the polling action did not result with a message (as opposed to 
+polling spans, where we want to capture such as well).
+
+### Typing
+
+- Transactions: 
+  - `transaction.type`: `messaging`
+- Spans: 
+  - `span.type`: `messaging` 
+  - `span.subtype`: the name of the framework - e.g. `jms`, `kafka`, `rabbitmq` 
+  - `span.action`: `send`, `receive` or `poll`
+  
+### Naming
+
+Transaction and span names may* follow this pattern: `<MSG-FRAMEWORK> SEND/RECEIVE/POLL to/from <QUEUE-NAME>`.
+Examples:
+- `JMS SEND to MyQueue`
+- `RabbitMQ RECEIVE from MyQueue`**
+- `RabbitMQ POLL from MyExchange`**
+
+Agents may deviate from this pattern, as long as they ensure a proper cardinality is maintained, that is- neither too low nor too high. 
+For example, agents may choose to name all transactions/spans reading-from/sending-to temporary queues equally. 
+On the other end, agents may choose to append a cardinality-increasing factor to the name, like the routing key in RabbitMQ.
+
+\* Java agent's instrumentation for Kafka does not follow this pattern at the moment.
+
+#### \** RabbitMQ naming specifics 
+
+In RabbitMQ, queues are only relevant in the receiver side, so the exchange name is used instead for sender spans.
+When the default exchange is used (denoted with an empty string), it should be replaced with `<default>`. 
+
+Agents may add an opt-in config to append the routing key to the name as well, for example: `RabbitMQ RECEIVE from MyExchange\58D7EA987`.
+
+For RabbitMQ transaction and polling spans, the queue name is used instead, whenever available (i.e. when the polling yields a message).
+
+### Context fields
+
+- **`context.message.queue.name`**: optional for `messaging` spans and transactions. Indexed as keyword. Wherever the broker terminology 
+uses "topic", this field will contain the topic name. In RabbitMQ, whenever the queue name is not available, use exchange name instead.
+- **`context.message.age.ms`**: optional for message/record receivers only (transactions or spans). 
+A numeric field indicating the message's age in milliseconds. Relevant for transactions and 
+`receive` spans that receive valid messages. There is no accurate definition as to how this is calculated. If the messaging framework 
+provides a timestamp for the message- agents may use it. Otherwise, the sending agent can add a timestamp _indicated as milliseconds since 
+epoch UTC_ to the message's metadata to be retrieved by the receiving agent. If a timestamp is not available- agents should omit this field. 
```

Comment:
What's the name of the metadata field agents should use? Does the Java agent already do that?

Reply:
Good point for interoperability! Currently the Java agent doesn't add such as all frameworks we support to date has a dedicated timestamp header/property. I will change the phrasing to indicate it can be a future option if we come across a framework that doesn't have such.